bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/handler/DanfossAirUnitHandler.java

```diff
@@ -107,54 +112,73 @@ public void initialize() {
-                    logger.debug("Cannot update channel {}: an error occurred retrieving the value: {}",
-                            channel.getChannelName(), e.getMessage());
-                }
-            }
+        logger.debug("Updating DanfossHRV data '{}'", getThing().getUID());
 
-            if (getThing().getStatus() == ThingStatus.OFFLINE) {
-                updateStatus(ThingStatus.ONLINE);
+        for (Channel channel : Channel.values()) {
+            if (Thread.interrupted()) {
+                logger.debug("Polling thread interrupted...");
+                return;
+            }
+            try {
+                updateState(channel.getGroup().getGroupName(), channel.getChannelName(),
+                        channel.getReadAccessor().access(localAirUnit));
+            } catch (UnexpectedResponseValueException e) {
+                updateState(channel.getGroup().getGroupName(), channel.getChannelName(), UnDefType.UNDEF);
+                logger.debug(
+                        "Cannot update channel {}: an unexpected or invalid response has been received from the air unit: {}",
+                        channel.getChannelName(), e.getMessage());
+            } catch (IOException e) {
+                updateState(channel.getGroup().getGroupName(), channel.getChannelName(), UnDefType.UNDEF);
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, e.getMessage());
+                logger.debug("Cannot update channel {}: an error occurred retrieving the value: {}",
+                        channel.getChannelName(), e.getMessage());
             }
         }
+
+        if (getThing().getStatus() == ThingStatus.OFFLINE) {
+            updateStatus(ThingStatus.ONLINE);
+        }
     }
 
     @Override
     public void dispose() {
         logger.debug("Disposing Danfoss HRV handler '{}'", getThing().getUID());
 
-        if (pollingJob != null) {
-            pollingJob.cancel(true);
-            pollingJob = null;
+        stopPolling();
+
+        this.airUnit = null;
+
+        DanfossAirUnitCommunicationController localCommunicationController = this.communicationController;
+        if (localCommunicationController != null) {
+            localCommunicationController.disconnect();
         }
+    }
 
-        if (hrv != null) {
-            hrv.cleanUp();
-            hrv = null;
+    private synchronized void startPolling() {
+        this.pollingJob = scheduler.scheduleWithFixedDelay(this::updateAllChannels, POLLING_INTERVAL_SECONDS,
+                config.refreshInterval, TimeUnit.SECONDS);
+    }
+
+    private synchronized void stopPolling() {
+        ScheduledFuture<?> localPollingJob = this.pollingJob;
+        if (localPollingJob != null) {
+            localPollingJob.cancel(true);
         }
+        this.pollingJob = null;
     }
 
     private void updateState(String groupId, String channelId, State state) {
```

Comment:
This should not happen, should it? So I'd either remove the check (fail early) or add a log message here, otherwise I think updates will get silently discarded (will be hard to track)

Reply:
Hmm, unsure what you are referring to, looks like line being commented is:
`this.pollingJob = null;`

But your comment could be towards
```
        ScheduledFuture<?> localPollingJob = this.pollingJob;
        if (localPollingJob != null) {
```
?

Removing the check will reintroduce this warning: Potential null pointer access: _The variable localPollingJob may be null at this location_

I'm not sure if it's worth logging, as it really shouldn't happen, unless dispose() is called twice or something, which would lead to more severe problems.

Reply:
I meant the null check yes. They are a real pain, especially when the tell the opposite of what my IDE (IntelliJ) does ;-)
So let's leave the check to keep the checker happy.